Update and rename 4.2_cpp_tutorial.md to 4.2_cpp_tutorial_en.md (#12)
qulacs/qulacs-rtd: en/source/intro/4.2_cpp_tutorial_en.md

}
```

> ### Obtain data of quantum state

Reviewer: ```suggestion
### Obtain data of quantum states
```

> The quantum state is expressed as an array of length $2^n$. Note that if the quantum state is formed by GPU, and If the $n$ is large, it can become extremely heavy operation.

Reviewer: ```suggestion
The quantum state is expressed as an array of length $2^n$. Especially when using GPU, note that,  if $n$ is large, it can become very computationally heavy.
```

	QuantumState state(n);
	state.set_zero_state();

	// If use GNU C++, obtain double_Complex array
	// If use MSVC, obtain std::complex<double> array
	const CTYPE* raw_data_c = state.data_c();

	// obtain std::complex<double> array
	const CPPCTYPE* raw_data_cpp = state.data_cpp();
}
```

If you want to set the quantum state directly to the specified array, it is recommended to create the corresponding quantum gate and perform it as an operation of the quantum gate.

> ## Quantum Gate

Reviewer: ```suggestion
## Quantum Gates
```

> ### Generation and operation of quantum gate

Reviewer: ```suggestion
### Generation and operation of quantum gates
```

- Noise : BitFlipNoise, DephasingNoise, IndepenedentXZNoise, DepolarizingNoise

> ### Merge of quantum gates

Reviewer: ```suggestion
### Combining quantum gates
```

}
```

> ### Special quantum gate and common quantum gate

Reviewer: ```suggestion
### Special quantum gates and common quantum gates
```

}
```

Please check the API documents for details of speacial quantum gate.

> ### Obtain the gate matrix of quantum gate

Reviewer: ```suggestion
### Obtaining the gate matrix of a quantum gate
```

```

> ### Obtain information of quantum gate

Reviewer: ```suggestion
### Obtaining information of a quantum gate
```

	return 0;
}
```

> ### Implement of common quantum gate

Reviewer: ```suggestion
### Implementation of common quantum gates
```

cppsim implements various maps of quantum information in the following forms.
#### Unitary operation

> Implemented as quantum gate.

Reviewer: ```suggestion
Implemented as quantum gates.
```

> #### Projection operator and Kraus operator, etc.

Reviewer: ```suggestion
#### Projection operators, Kraus operators, and etc.
```

> Implemented as quantum gate. In general, the norm of quantum state is not preserved after operation. The gate can be generated by <code>DenseMatrix</code>.

Reviewer: ```suggestion
Implemented as quantum gates. In general, the norm of a quantum state is not preserved after an operation. The gate can be generated by <code>DenseMatrix</code>.
```

```

#### CP-map
If Kraus-rank is 1, please treat it as a single Kraus operator as described above. In other cases, please adjust the Kraus operator so that it becomes TP, and then adjust it by applying the <code>Identity</code> operator multiplied by a constant with the <code>multiply_scalar</code> function.

> ## Quantum circuit

Reviewer: ```suggestion
## Quantum circuits
```

> ### Constitution of quantum circuit

Reviewer: ```suggestion
### Constitution of quantum circuits
```

	QuantumState state(n);
	state.set_zero_state();

	// Define quantum circuit
	QuantumCircuit circuit(n);

	// Add gate to quantum circuit
	for(int i=0;i<n;++i){
		circuit.add_H_gate(i);
	}

	// Gate defined by user can also be added
	for(int i=0;i<n;++i){
		circuit.add_gate(gate::H(i));
	}

	// Operate quantum circuit to state
	circuit.update_quantum_state(&state);
	return 0;
}
```
Note: the quantum circuit added by <code>add_gate</code> is released from memory when the quantum circuit is released. Therefore, the assigned gate cannot be reused. If you want to reuse the gate given as an argument, make a copy of itself using gate.copy or use the <code>add_gate_copy</code> function. But in this case, you have to release the gate by yourself.

> ### Optimization of quantum circuit

Reviewer: ```suggestion
### Optimization of quantum circuits
```

}
```

> ### Obtain debug information of quantum circuit

Reviewer: ```suggestion
### Obtaining debug information of quantum circuits
```

		}
	}

	// Output information of quantum circuit
	std::cout << circuit << std::endl;
	return 0;
}
```

> ## Observable

Reviewer: ```suggestion
## Observables
```

> ### Create observable

Reviewer: ```suggestion
### Creating observables
```
